### ISSUE TITLE
Bug: no-unnecessary-condition false positive with await

### ISSUE THREAD

**Issue by lydell** [typescript-eslint/typescript-eslint#10143]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=IYZwngdgxgBAZgV2gFwJYHsIwE5IBQCUMA3gFAwwA2ApsjAO6gDKl69MAvPMJSNQNzkYUTCDpoAttXQJkASQAmnGH2QAVVFJnI8hTgD4SQioxAs2y5LgHHho9DQB0rAOZ4ARGoAWqEDF8wyMAA1qgQLjDADD40gGQE7gSCFAC%2BADQwAMwADDmJQiIQYjjUIAAOotTKwIyodHC0UF4eIOhSyD7hCUnCNMDYGlqyeJLSsop5FAD0kzAQ6AC0SBDUUCUgfWDzBQq1GFgBc3RewKWlYDBe1NjUAFxCqHAweKbm9ERkFBQFLU6uHrEAI1kswstQA5H4FJhqABCLpCZL5ey-dBua5lCp5RGkRAoPY4JAAJj0HyotAYzFY7C4cB4fG633EmlG8iUXFUAxZuiIHEMpJMlIsXCsCBsnzshQc1GcqI83gCASCoXCkWiqDi8JS6WyuW69WQjWarVoHRcCUccDCPEoYG5Bh61D6nO0w2Z2nGBEc7WoEDw6PKhUqvKM4umMCBdFq-j8x1O5zCgR8fmqwFq8zg10qADcriA9jDbA8ni8qe9bF9kdK-u4AELAubsWqATAJIdC4RNPojxd8pTK0SUA3wO8ksUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
async function run() {
  let wasSlow = false;
  const timeoutId = setTimeout(() => {
    wasSlow = true;
    console.log("This is taking a while…");
  }, 3000);
  const response = await fetch("something");
  clearTimeout(timeoutId);
  // no-unnecessary-condition is not happy here:
  if (wasSlow) {
    console.log("…but now it's done!");
  }
  console.log(response);
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

`no-unnecessary-condition` should _not_ report anything.

### Actual Result

This line:

```ts
  if (wasSlow) {
```

gets this error:

```
Unnecessary conditional, value is always falsy. 9:7 - 9:14
```

### Additional Info

This does not happen when not using `await`:

```ts
function run2() {
  let wasSlow = false;
  const timeoutId = setTimeout(() => {
    wasSlow = true;
    console.log("This is taking a while…");
  }, 3000);
  return fetch("something").finally(() => clearTimeout(timeoutId)).then(response => {
    // but it is happy in this await-free version!
    if (wasSlow) {
      console.log("But now it’s done!");
    }
    console.log(response);
  });
}
```

**Comment by bradzacher** [comment]
https://typescript-eslint.io/rules/no-unnecessary-condition/#when-not-to-use-it


[playground](https://typescript-eslint.io/play#ts=5.5.2&fileType=.ts&code=IYZwngdgxgBAZgV2gFwJYHsIwE5IBQCUMA3gFAwwA2ApsjAO6gDKl69MAvPMJSNQNzkYfZABVUAW2roEyPIU4A%2BEkIqMQLNpxjJcA1TCiYQ6GgDpWAczwAiUQAtUIGE53AA1qgiWYwBo5pAMgIbAkEKAF8AGhgAZgAGBNChdU16MJgAegyYAD0AfiEjCBA6bGoQAAdjam1gRlQ6OFooe1sTKWRHbxD0rJgIdABaJAhqKHKQYGwwQaKAEwaMLFcBuntgCoqwGHtqMoAuIVQ4GDwU1noiMgoKPtyCm8NjU2oLdGsbQIAjWX6tBoA5M45phqABCHpCcKFZ7mKx4MqVapJcJAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by lydell** [comment]
Thanks! I actually did read through the entire documentation, but didn’t make the connection between that paragraph and the issue I was facing.
